src/main/java/bdv/fx/viewer/ViewerPanelFX.java

```diff
@@ -570,32 +570,49 @@ public OverlayPane getDisplay()
 		return range;
 	}
 
-	private static GridPane makeGrid(final RenderUnit.ImagePropertyGrid grid)
+	private CanvasPane makeCanvas(final RenderUnit.ImagePropertyGrid grid)
 	{
-		final GridPane pane = new GridPane();
-
+		final CanvasPane canvasPane = new CanvasPane(1, 1);
 		if (grid == null)
-			return pane;
+			return canvasPane;
 
 		final CellGrid cellGrid = grid.getGrid();
-		pane.setMinWidth(1);
-		pane.setMinHeight(1);
-		pane.setPrefWidth(cellGrid.getImgDimensions()[0]);
-		pane.setPrefHeight(cellGrid.getImgDimensions()[0]);
 		final long[] gridPos = new long[2];
-		final long[] cellMin = new long[2];
-		final int[] cellDims = new int[2];
-		final int numTiles = grid.numTiles();
-		for (int i = 0; i < numTiles; ++i) {
+		for (int i = 0; i < grid.numTiles(); ++i) {
+			final long[] cellMin = new long[2];
+			final int[] cellDims = new int[2], upscaledCellDims = new int[2];
 			cellGrid.getCellGridPositionFlat(i, gridPos);
 			cellGrid.getCellDimensions(gridPos, cellMin, cellDims);
-			final ReadOnlyObjectProperty<Image> image = grid.imagePropertyAt(i);
-			final ImagePane imagePane = new ImagePane(cellDims[0], cellDims[1]);
-			imagePane.imageProperty().bind(image);
-			LOG.debug("Putting render block {} into cell at position {}", i, gridPos);
-			pane.add(imagePane, (int) gridPos[0], (int) gridPos[1]);
+
+			final int[] imageSize = new int[2], unpaddedImageSize = new int[2];
+			final ReadOnlyObjectProperty<RenderUnit.RenderedImage> renderedImageProperty = grid.renderedImagePropertyAt(i);
+			renderedImageProperty.addListener((obs, oldv, newv) -> {
+				if (newv == null) {
+					canvasPane.getCanvas().getGraphicsContext2D().setFill(Color.BLACK);
```

Comment:
Please add comment that `newv == null` check and filling with black color is only necessary because there is no `src` `BlendMode` in JavaFX, also describe the `src` blend mode like this:
> Overwrites the destination with the source, regardless of alpha

Paste these links into the comment as well:
https://bugs.openjdk.java.net/browse/JDK-8092156
https://books.google.com/books?id=UxM2iXiFqbMC&pg=PA97&lpg=PA97&dq=Porter-Duff+%22source+over+destination%22&source=bl&ots=hYj5U2X5Lk&sig=ACfU3U095X67T4FghqCpgYrAhEWbNtd_xQ&hl=en&sa=X&ved=2ahUKEwjyg4rRufXfAhXvuFkKHfK8BysQ6AEwAnoECAcQAQ#v=snippet&q=%22overwrites%20the%20destination%20with%20the%20source%2C%20regardless%20of%20alpha%22&f=false
https://docs.oracle.com/javase/8/javafx/api/javafx/scene/effect/BlendMode.html